### ISSUE TITLE
Multi threading the Document Rendering

### ISSUE THREAD

**Issue by ajos6183** [empira/MigraDoc#11]
Hi @ThomasHoevel / @StLange 

This is only a question. I am using the following code to render the PDF document and at the same time monitor the progress. I have removed the progress monitoring code for simplicity.

        Dim pdfRenderer = New PdfDocumentRenderer(True) With {.Document = report}
        pdfRenderer.PrepareRenderPages()
        For i As Integer = 1 To pdfRenderer.DocumentRenderer.FormattedDocument.PageCount
            Dim page = pdfRenderer.PdfDocument.AddPage()
            Dim pageInfo As PageInfo = pdfRenderer.DocumentRenderer.FormattedDocument.GetPageInfo(i)
            page.Width = pageInfo.Width
            page.Height = pageInfo.Height
            page.Orientation = pageInfo.Orientation
            Using gfx As PdfSharp.Drawing.XGraphics = PdfSharp.Drawing.XGraphics.FromPdfPage(page)
                gfx.MUH = If(pdfRenderer.Unicode, PdfFontEncoding.Unicode, PdfFontEncoding.WinAnsi)
                pdfRenderer.DocumentRenderer.RenderPage(gfx, i)
            End Using
        Next
        pdfRenderer.PdfDocument.PageLayout = PdfSharp.Pdf.PdfPageLayout.SinglePage

The PDF document is around 3500 pages. It takes about 130 seconds to prepare the pages and another 100 seconds to render. Is there a way i can incorporate multi threading to speed things up? My simplistic attempts to do that for the rendering have failed.

Thanks,
Anthony

**Comment by ajos6183** [comment]
Is anyone able to help with this?

**Comment by TH-Soft** [comment]
Anyone now? The original question is aimed at two persons only.

Not my area of expertise. No definite answer, just some guesses.
Currently a new section starts on a new page. It could be possible to use multi-threading for the layout process using one thread for each section.
It should be possible to render multi-threaded using one thread per page, but it may be necessary to add pages single-threaded (using a semaphore).
I don't know if rendering is thread-safe. This may require changes in MigraDoc or PDFsharp.
There are some locks in the font management of PDFsharp, but multi-threading was not yet tested intensively.
You ask if you can do it. I cannot judge your abilities. I'm sure it can be done in general. I'm not sure if I can do it. I cannot guess how long it would take me.

**Comment by ajos6183** [comment]
Thank you for your help. I appreciate it. I will try your suggestions.

**Comment by TH-Soft** [comment]
Things are a bit more complicated than I realised one hour ago.
I guess you have to serialise all calls that lead to new objects in the PDF file.

Pages: You can easily serialise these calls in your code: call AddPage() for all pages before starting the first render thread.

Fonts: There are unverified reports that the locking mechanism for fonts in PDFsharp was not working. Known workaround: Use a single thread to create a dummy PDF that uses all the fonts you will ever need.
If your MigraDoc document uses e.g. Arial, Courier, and Times, then create a page with text in Arial Regular, Arial Bold, Arial Italic, Arial Bold Italic, Courier Regular, Courier Bold and so forth.
It might be worth using the workaround even if it is not needed for fonts as this might simplify dealing with other object types.

Images: I would tackle this by adding a Lock to any DrawImage call in the XGraphics class.

Other objects: There could be more objects I do not currently think of.  Use the same lock you also use for images if there are any other objects.

As always: It could be simpler to get it working for specific MigraDoc documents that use limited features than making a general implementation that works for all MigraDoc features.

**Comment by ajos6183** [comment]
Thanks for the follow up comments. When i get to it, i will research what you explained and try to implement it.
